README.md
Prometheus PuppetDB exporter
============================

## Usage

```
Usage:
  prometheus-puppetdb-exporter [OPTIONS]

Application Options:
      --version          Show version.
  -u, --puppetdb-url=    PuppetDB base URL. (default: https://puppetdb:8081/pdb/query) [$PUPPETDB_URL]
      --cert-file=       A PEM encoded certificate file. [$PUPPETDB_CERT_FILE]
      --key-file=        A PEM encoded private key file. [$PUPPETDB_KEY_FILE]
      --ca-file=         A PEM encoded CA's certificate. [$PUPPETDB_CA_FILE]
      --ssl-skip-verify  Skip SSL verification. [$PUPPETDB_SSL_SKIP_VERIFY]
      --scrape-interval= Duration between two scrapes. (default: 5s) [$PUPPETDB_SCRAPE_INTERVAL]
      --listen-address=  Address to listen on for web interface and telemetry. (default: 0.0.0.0:9635)
                         [$PUPPETDB_LISTEN_ADDRESS]
      --metric-path=     Path under which to expose metrics. (default: /metrics) [$PUPPETDB_METRIC_PATH]
      --verbose          Enable debug mode [$PUPPETDB_VERBOSE]
      --unreported-node= Tag nodes as unreported if the latest report is older than the defined duration.
                         (default: 2h) [$PUPPETDB_UNREPORTED_NODE]
      --categories=      Report metrics categories to scrape. (default: resources,time,changes,events)
                         [$REPORT_METRICS_CATEGORIES]

Help Options:
  -h, --help             Show this help message
```

## Metrics

```
# HELP puppetdb_exporter_build_info puppetdb exporter build informations
# TYPE puppetdb_exporter_build_info gauge
puppetdb_exporter_build_info{build_date="2019-02-18",commit_sha="XXXXXXXXXX",golang_version="go1.11.4",version="1.0.0"} 1
# HELP puppetdb_node_report_status_count Total count of reports status by type
# TYPE puppetdb_node_report_status_count gauge
puppetdb_node_report_status_count{status="changed"} 1
puppetdb_node_report_status_count{status="failed"} 1
puppetdb_node_report_status_count{status="unchanged"} 1
```


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/camptocamp/prometheus-puppetdb-exporter/tree/1.1.0) (2020-12-03)

[Full Changelog](https://github.com/camptocamp/prometheus-puppetdb-exporter/compare/1.0.0...1.1.0)

**Closed issues:**

- Deactivated nodes still appear [\#8](https://github.com/camptocamp/prometheus-puppetdb-exporter/issues/8)

**Improvements:**

- Remove legacy `vendor/` directory

## [1.0.0](https://github.com/camptocamp/prometheus-puppetdb-exporter/tree/1.0.0) (2020-06-09)

[Full Changelog](https://github.com/camptocamp/prometheus-puppetdb-exporter/compare/0.1.0...1.0.0)

**Breaking changes:**

- Switch port to 9635 [\#5](https://github.com/camptocamp/prometheus-puppetdb-exporter/pull/5) ([bastelfreak](https://github.com/bastelfreak))

## [0.1.0](https://github.com/camptocamp/prometheus-puppetdb-exporter/tree/0.1.0) (2020-06-08)

[Full Changelog](https://github.com/camptocamp/prometheus-puppetdb-exporter/compare/8499b362f2f346f1ce58a60d5299d0de628556aa...0.1.0)

**Closed issues:**

- No nodes exist? [\#3](https://github.com/camptocamp/prometheus-puppetdb-exporter/issues/3)
- Report nodes as "unreported" if last report is more that 2 hours old [\#1](https://github.com/camptocamp/prometheus-puppetdb-exporter/issues/1)

**Merged pull requests:**

- Allow to select report metrics categories to scrape [\#2](https://github.com/camptocamp/prometheus-puppetdb-exporter/pull/2) ([mcanevet](https://github.com/mcanevet))



\* *This Changelog was automatically generated by [github_changelog_generator](https://github.com/github-changelog-generator/github-changelog-generator)*


internal/exporter/exporter.go
package exporter

import (
	"fmt"
	"strings"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	log "github.com/sirupsen/logrus"

	"github.com/camptocamp/prometheus-puppetdb-exporter/internal/puppetdb"
)

// Exporter implements the prometheus.Exporter interface, and exports PuppetDB metrics
type Exporter struct {
	client    *puppetdb.PuppetDB
	namespace string
	metrics   map[string]*prometheus.GaugeVec
}

var (
	metricMap = map[string]string{
		"node_status_count": "node_status_count",
	}
)

// NewPuppetDBExporter returns a new exporter of PuppetDB metrics.
func NewPuppetDBExporter(url, certPath, caPath, keyPath string, sslSkipVerify bool, categories map[string]struct{}) (e *Exporter, err error) {
	e = &Exporter{
		namespace: "puppetdb",
	}

	opts := &puppetdb.Options{
		URL:        url,
		CertPath:   certPath,
		CACertPath: caPath,
		KeyPath:    keyPath,
		SSLVerify:  sslSkipVerify,
	}

	e.client, err = puppetdb.NewClient(opts)
	if err != nil {
		log.Fatalf("failed to create new client: %s", err)
		return
	}

	e.initGauges(categories)

	return
}

// Describe outputs PuppetDB metric descriptions
func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
	for _, m := range e.metrics {
		m.Describe(ch)
	}
}

// Collect fetches new metrics from the PuppetDB and updates the appropriate metrics
func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
	for _, m := range e.metrics {
		m.Collect(ch)
	}
}

// Scrape scrapes PuppetDB and update metrics
func (e *Exporter) Scrape(interval time.Duration, unreportedNode string, categories map[string]struct{}) {
	var statuses map[string]int

	unreportedDuration, err := time.ParseDuration(unreportedNode)
	if err != nil {
		log.Errorf("failed to parse unreported duration: %s", err)
		return
	}

	for {
		statuses = make(map[string]int)

		nodes, err := e.client.Nodes()
		if err != nil {
			log.Errorf("failed to get nodes: %s", err)
		}

		e.metrics["report"].Reset()
		e.metrics["node_report_status_count"].Reset()

		for _, node := range nodes {
			var deactivated string
			if node.Deactivated == "" {
				deactivated = "false"
			} else {
				deactivated = "true"
			}

			if node.ReportTimestamp == "" {
				if deactivated == "false" {
					statuses["unreported"]++
				}
				continue
			}
			latestReport, err := time.Parse("2006-01-02T15:04:05Z", node.ReportTimestamp)
			if err != nil {
				if deactivated == "false" {
					statuses["unreported"]++
				}
				log.Errorf("failed to parse report timestamp: %s", err)
				continue
			}
			e.metrics["report"].With(prometheus.Labels{"environment": node.ReportEnvironment, "host": node.Certname, "deactivated": deactivated}).Set(float64(latestReport.Unix()))

			if deactivated == "false" {
				if latestReport.Add(unreportedDuration).Before(time.Now()) {
					statuses["unreported"]++
				} else if node.LatestReportStatus == "" {
					statuses["unreported"]++
				} else {
					statuses[node.LatestReportStatus]++
				}
			}

			if node.LatestReportHash != "" {
				reportMetrics, _ := e.client.ReportMetrics(node.LatestReportHash)
				for _, reportMetric := range reportMetrics {
					_, ok := categories[reportMetric.Category]
					if ok {
						category := fmt.Sprintf("report_%s", reportMetric.Category)
						e.metrics[category].With(prometheus.Labels{"name": strings.ReplaceAll(strings.Title(reportMetric.Name), "_", " "), "environment": node.ReportEnvironment, "host": node.Certname}).Set(reportMetric.Value)
					}
				}
			}
		}

		for statusName, statusValue := range statuses {
			e.metrics["node_report_status_count"].With(prometheus.Labels{"status": statusName}).Set(float64(statusValue))
		}

		time.Sleep(interval)
	}
}

func (e *Exporter) initGauges(categories map[string]struct{}) {
	e.metrics = map[string]*prometheus.GaugeVec{}

	e.metrics["node_report_status_count"] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: e.namespace,
		Name:      "node_report_status_count",
		Help:      "Total count of reports status by type",
	}, []string{"status"})

	for category := range categories {
		metricName := fmt.Sprintf("report_%s", category)
		e.metrics[metricName] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
			Namespace: "puppet",
			Name:      metricName,
			Help:      fmt.Sprintf("Total count of %s per status", category),
		}, []string{"name", "environment", "host"})

	}

	e.metrics["report"] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Namespace: "puppet",
		Name:      "report",
		Help:      "Timestamp of latest report",
	}, []string{"environment", "host", "deactivated"})

	for _, m := range e.metrics {
		prometheus.MustRegister(m)
	}
}


internal/puppetdb/puppetdb.go
package puppetdb

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
)

// PuppetDB stores informations used to connect to a PuppetDB
type PuppetDB struct {
	options *Options
	client  *http.Client
}

// Options contains the options used to connect to a PuppetDB
type Options struct {
	URL        string
	CertPath   string
	CACertPath string
	KeyPath    string
	SSLVerify  bool
}

// Node is a structure returned by a PuppetDB
type Node struct {
	Certname           string `json:"certname"`
	Deactivated        string `json:"deactivated"`
	LatestReportStatus string `json:"latest_report_status"`
	ReportEnvironment  string `json:"report_environment"`
	ReportTimestamp    string `json:"report_timestamp"`
	LatestReportHash   string `json:"latest_report_hash"`
}

// ReportMetric is a structure returned by a PuppetDB
type ReportMetric struct {
	Name     string  `json:"name"`
	Value    float64 `json:"value"`
	Category string  `json:"category"`
}

// NewClient creates a new PuppetDB client
func NewClient(options *Options) (p *PuppetDB, err error) {
	var transport *http.Transport

	puppetdbURL, err := url.Parse(options.URL)
	if err != nil {
		err = fmt.Errorf("failed to parse PuppetDB URL: %v", err)
		return
	}

	if puppetdbURL.Scheme != "http" && puppetdbURL.Scheme != "https" {
		err = fmt.Errorf("%s is not a valid http scheme", puppetdbURL.Scheme)
		return
	}

	if puppetdbURL.Scheme == "https" {
		// Load client cert
		cert, err := tls.LoadX509KeyPair(options.CertPath, options.KeyPath)
		if err != nil {
			err = fmt.Errorf("failed to load keypair: %s", err)
			return nil, err
		}

		// Load CA cert
		caCert, err := ioutil.ReadFile(options.CACertPath)
		if err != nil {
			err = fmt.Errorf("failed to load ca certificate: %s", err)
			return nil, err
		}
		caCertPool := x509.NewCertPool()
		caCertPool.AppendCertsFromPEM(caCert)

		// Setup HTTPS client
		tlsConfig := &tls.Config{
			Certificates:       []tls.Certificate{cert},
			RootCAs:            caCertPool,
			InsecureSkipVerify: !options.SSLVerify,
		}
		tlsConfig.BuildNameToCertificate()
		transport = &http.Transport{TLSClientConfig: tlsConfig}
	} else {
		transport = &http.Transport{}
	}

	p = &PuppetDB{
		client:  &http.Client{Transport: transport},
		options: options,
	}
	return
}

// Nodes returns the list of nodes
func (p *PuppetDB) Nodes() (nodes []Node, err error) {
	err = p.get("nodes", "[\"or\", [\"=\", [\"node\", \"active\"], false], [\"=\", [\"node\", \"active\"], true]]", &nodes)
	if err != nil {
		err = fmt.Errorf("failed to get nodes: %s", err)
		return
	}
	return
}

// ReportMetrics returns the list of reportMetrics
func (p *PuppetDB) ReportMetrics(reportHash string) (reportMetrics []ReportMetric, err error) {
	err = p.get(fmt.Sprintf("reports/%s/metrics", reportHash), "", &reportMetrics)
	if err != nil {
		err = fmt.Errorf("failed to get reports: %s", err)
		return
	}
	return
}

func (p *PuppetDB) get(endpoint string, query string, object interface{}) (err error) {
	base := strings.TrimRight(p.options.URL, "/")
	var myurl string
	if query == "" {
		myurl = fmt.Sprintf("%s/v4/%s", base, endpoint)
	} else {
		myurl = fmt.Sprintf("%s/v4/%s?query=%s", base, endpoint, url.QueryEscape(query))
	}
	req, err := http.NewRequest("GET", myurl, strings.NewReader(""))
	if err != nil {
		err = fmt.Errorf("failed to build request: %s", err)
		return
	}
	resp, err := p.client.Do(req)
	if err != nil {
		err = fmt.Errorf("failed to call API: %s", err)
		return
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		err = fmt.Errorf("failed to read response: %s", err)
		return
	}
	err = json.Unmarshal(body, object)
	if err != nil {
		err = fmt.Errorf("failed to unmarshal: %s", err)
		return
	}
	return
}


main.go
package main

import (
	"net/http"
	"os"
	"runtime"
	"strings"
	"time"

	"github.com/jessevdk/go-flags"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	log "github.com/sirupsen/logrus"

	"github.com/camptocamp/prometheus-puppetdb-exporter/internal/exporter"
)

// Config stores handler's configuration
type Config struct {
	Version        bool   `long:"version" description:"Show version."`
	PuppetDBUrl    string `short:"u" long:"puppetdb-url" description:"PuppetDB base URL." env:"PUPPETDB_URL" required:"true" default:"https://puppetdb:8081/pdb/query"`
	CertFile       string `long:"cert-file" description:"A PEM encoded certificate file." env:"PUPPETDB_CERT_FILE"`
	KeyFile        string `long:"key-file" description:"A PEM encoded private key file." env:"PUPPETDB_KEY_FILE"`
	CACertFile     string `long:"ca-file" description:"A PEM encoded CA's certificate." env:"PUPPETDB_CA_FILE"`
	SSLSkipVerify  bool   `long:"ssl-skip-verify" description:"Skip SSL verification." env:"PUPPETDB_SSL_SKIP_VERIFY"`
	ScrapeInterval string `long:"scrape-interval" description:"Duration between two scrapes." env:"PUPPETDB_SCRAPE_INTERVAL" default:"5s"`
	ListenAddress  string `long:"listen-address" description:"Address to listen on for web interface and telemetry." env:"PUPPETDB_LISTEN_ADDRESS" default:"0.0.0.0:9635"`
	MetricPath     string `long:"metric-path" description:"Path under which to expose metrics." env:"PUPPETDB_METRIC_PATH" default:"/metrics"`
	Verbose        bool   `long:"verbose" description:"Enable debug mode" env:"PUPPETDB_VERBOSE"`
	UnreportedNode string `long:"unreported-node" description:"Tag nodes as unreported if the latest report is older than the defined duration." env:"PUPPETDB_UNREPORTED_NODE" default:"2h"`
	Categories     string `long:"categories" description:"Report metrics categories to scrape." env:"REPORT_METRICS_CATEGORIES" default:"resources,time,changes,events"`
}

var (
	// VERSION, BUILD_DATE, GIT_COMMIT are filled in by the build script
	version    = "<<< filled in by build >>>"
	buildDate  = "<<< filled in by build >>>"
	commitSha1 = "<<< filled in by build >>>"
)

func main() {
	var c Config
	parser := flags.NewParser(&c, flags.Default)
	if _, err := parser.Parse(); err != nil {
		if flagsErr, ok := err.(*flags.Error); ok && flagsErr.Type == flags.ErrHelp {
			os.Exit(0)
		} else {
			os.Exit(1)
		}
	}

	log.Printf("PuppetDB Metrics Exporter %s    build date: %s    sha1: %s    Go: %s",
		version, buildDate, commitSha1,
		runtime.Version(),
	)
	if c.Verbose {
		log.SetLevel(log.DebugLevel)
		log.Debugln("Enabling debug output")
	} else {
		log.SetLevel(log.InfoLevel)
	}

	if c.Version {
		return
	}

	interval, err := time.ParseDuration(c.ScrapeInterval)
	if err != nil {
		log.Fatalf("failed to parse scrape interval duration: %s", err)
	}

	// Create a map[string]struct{} of categories to provide an efficient way to
	// find if a category exists in the list of categories.
	cats := strings.Split(c.Categories, ",")
	categories := make(map[string]struct{}, len(cats))
	for _, category := range cats {
		categories[category] = struct{}{}
	}
	exp, err := exporter.NewPuppetDBExporter(c.PuppetDBUrl, c.CertFile, c.CACertFile, c.KeyFile, c.SSLSkipVerify, categories)
	if err != nil {
		log.Fatalf("failed to initialize exporter: %s", err)
	}

	go exp.Scrape(interval, c.UnreportedNode, categories)

	buildInfo := prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "puppetdb_exporter_build_info",
		Help: "puppetdb exporter build informations",
	}, []string{"version", "commit_sha", "build_date", "golang_version"})
	buildInfo.WithLabelValues(version, commitSha1, buildDate, runtime.Version()).Set(1)
	prometheus.MustRegister(buildInfo)

	http.Handle(c.MetricPath, promhttp.Handler())
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`
<html>
<head><title>Prometheus PuppetDB Exporter v` + version + `</title></head>
<body>
<h1>Prometheus PuppetDB Exporter ` + version + `</h1>
<p><a href='` + c.MetricPath + `'>Metrics</a></p>
</body>
</html>
						`))
	})

	log.Infof("Providing metrics at %s%s", c.ListenAddress, c.MetricPath)
	log.Fatal(http.ListenAndServe(c.ListenAddress, nil))
}
